### DISCUSSION TITLE
Autofixture fixture setup questions

### DISCUSSION THREAD

**Discussion by rathoregee** [AutoFixture/AutoFixture#1291] [Q&A]
**My plan is setup kind of a builder pattern for S3, to perform various ops (put,get,list,delete etc...)**

```cs
using Amazon.S3;
using Amazon.S3.Model;
using AutoFixture;
using AutoFixture.AutoMoq;
using Moq;
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading;

namespace PdfUploder.Tests.UnitTests
{
    public class StorageBuilder
    {
        private IFixture _fixture;
        private readonly Mock<AmazonS3Client> _client;
        public StorageBuilder()
        {
            _fixture = new Fixture().Customize(new AutoMoqCustomization());
            _client = _fixture.Freeze<Mock<AmazonS3Client>>();
        }
        
        public StorageBuilder WithPut()
        {
            var res = _fixture.Create<PutObjectResponse>();
            Func<PutObjectRequest, CancellationToken, PutObjectResponse> response
                = (p, c) => { return res; };

               _client
              .Setup(x => x.PutObjectAsync(
                 It.IsAny<PutObjectRequest>(),
                 It.IsAny<CancellationToken>()))
              .ReturnsAsync(response);
            return this;
        }      

        public Mock<AmazonS3Client> Build()
        {
            return _client;
        }
    }
}
```

Q1.  when i run this command
` var res = _fixture.Create<PutObjectResponse>();`
if get this error
```
AutoFixture.ObjectCreationExceptionWithPath: 'AutoFixture was unable to create an instance from Amazon.S3.RequestCharged because creation unexpectedly failed with exception. Please refer to the inner exception to investigate the root cause of the failure
```

Q2. how to get fully prepared object of  Microsoft.AspNetCore.Http -->IFormFile
using this attribute, getting an object 
`public AutoMoqDataAttribute() : base(() => new Fixture().Customize(new AutoMoqCustomization() { ConfigureMembers = true })) {  }`

FileName
ContentType
ContentDisposition

but everything is coming null

**Comment by Ergamon** [comment]
Some Apis dont play nice with AutoFixture.

So your first problem is that you want to create an object of PutObjectResponse. The way AutoFixture is designed is that it tries to build an object graph for you.

So in a simple way you can think of it like this:
- Call constructor with every parameter created by AutoFixture
- Call setter of every property with a value created by AutoFixture.

I have no clue about the Amazon Api, but I took a quick look at the PutObjectResponse. This one has a property called RequestChanged. So what happens is that AutoFixture tries to create an instance of a RequestChanged object and fails as this class has no public constructor.

There are multiple ways to solve it, e.g. by telling AutoFixture how to create one, but the simplest way is to create one without this property.

```cs
var response = fixture.Build<PutObjectResponse>()
    .Without(r => r.RequestCharged)
    .Create();
```

For your second question I need more input.

Normally if you work with xUnit you create an attribute with your mocking framework of choice. 

Then you create tests like this:

```cs
[Theory]
[YourAttribute]
public void MyTest(IFormFile formFile)
{
  formFile.FileName.Should().NotBeNullOrEmpty();
}
```

The whole point of these attributes is to avoid repetitive code and increase readability.

Behind the scense nothing more than
```cs
fixture.Create<IFormFile>();
```
happens

**Comment by rathoregee** [comment]
thanks for reply - i got the idea

**Comment by rathoregee** [comment]
I am getting moq exception

```cs
 public class HttpClientSpecimenBuilder : ISpecimenBuilder
    {
        private MockEventBus _mockEventBusClient;
        private Mokta _mokta;
      
        public object Create(object request, ISpecimenContext context)
        {          
            _mockEventBusClient = context.Create<MockEventBus>();
            _mokta = context.Create<Mokta>();

            return CreateHttpClient();
        }

 private HttpClient CreateHttpClient()
        {
            var builder = new WebHostBuilder()
                .ConfigureTestServices(services =>
                {
                    services.AddSingleton(_mokta.OktaClient);
                    services.AddSingleton(_mockEventBusClient.EventBusClient);
                })
                .UseConfiguration(new ConfigurationBuilder().AddInMemoryCollection(new Dictionary<string, string>
                {
                    {
                        "OKTA_PRIVATE_KEY",
                        "mykey00000000000000000000000000000000000000"
                    }

                }).Build())
                .UseStartup<Startup>();

            var testServer = new TestServer(builder);
            return testServer.CreateClient();
        }




public static class FixtureHelper
    {
        public static IFixture CreateFixture()
        {
            var fixture = new Fixture();

            fixture.Customize(new AutoMoqCustomization
            {
                GenerateDelegates = true,
                ConfigureMembers = true
            });

            fixture.Customizations.Add(new HttpClientSpecimenBuilder());
            fixture.Customizations.Add(new OktaUserSpecimenBuilder());
            return fixture;
        }
    }


[AttributeUsage(AttributeTargets.Method, AllowMultiple = false)]
    public class AutoMoqDataAttribute : AutoDataAttribute
    {
        public AutoMoqDataAttribute() : base(FixtureHelper.CreateFixture) { }
    }


/////////////////////////////// unit test /////////////////////////////

 [Theory, AutoMoqData]
        public async Task CreateUser([Frozen] Mokta mokta,
            [Frozen] MockEventBus mockEventBusClient,
            HttpClient client,
            User user,
            SignUpRequest signUpRequest)
        {
            user.Profile.Email = Email;

            mokta.SetUpCreateUserResponse(user);

            mockEventBusClient.SetUpResponse<CreatePersonMessage, string>("person:create", HttpStatusCode.Created, NewPersonId);
            mockEventBusClient.SetUpResponse<CreateContactInfoMessage, string>("contactinfo:create", HttpStatusCode.Created, NewPersonId);
            mockEventBusClient.SetUpResponse<CreateUserMessage, string>("user:create", HttpStatusCode.Created, NewPersonId);

            var payload = new StringContent(JsonConvert.SerializeObject(signUpRequest),
                Encoding.UTF8, "application/json");

            var response = await client.PostAsync("users", payload);

            mokta.VerifyCreateUser(options => options.Profile.Email == Email);
            mockEventBusClient.Verify<CreatePersonMessage, string>("person:create", createPersonMessage => createPersonMessage.ForeName == FirstName && createPersonMessage.LastName == LastName);
            mockEventBusClient.Verify<CreateContactInfoMessage, string>("contactinfo:create", createContactInfoMessage =>
                    createContactInfoMessage.Type == "phone" &&
                    createContactInfoMessage.Label == "Contact Number" &&
                    createContactInfoMessage.OwnerType == "person" &&
                    createContactInfoMessage.OwnerId == NewPersonId &&
                    createContactInfoMessage.ContactValue == JsonConvert.SerializeObject(new
                    {
                        Number = ContactNumber
                    })
                );
            mockEventBusClient.Verify<CreateContactInfoMessage, string>("contactinfo:create", createContactInfoMessage =>
                    createContactInfoMessage.Type == "email" &&
                    createContactInfoMessage.Label == "Email" &&
                    createContactInfoMessage.OwnerType == "person" &&
                    createContactInfoMessage.OwnerId == NewPersonId &&
                    !string.IsNullOrEmpty(createContactInfoMessage.RelationshipId) &&
                    createContactInfoMessage.ContactValue == JsonConvert.SerializeObject(new
                    {
                        Email
                    })
                );
            mockEventBusClient.Verify<CreateUserMessage, string>("user:create", createUserMessage =>
                    createUserMessage.Id == NewOktaId &&
                    createUserMessage.PersonId == NewPersonId
                );

            Assert.Equal(HttpStatusCode.Created, response.StatusCode);
        }
```

the error i am getting this

Message: 
```
    Moq.MockException : 
    Expected invocation on the mock at least once, but was never performed: x => x.CreateUserAsync(It.Is<CreateUserWithoutCredentialsOptions>(options => options.Profile.Email == "name@email.com"), It.IsAny<CancellationToken>())
```    

**I need help to setup setup part**
I feel main problem is to share single object and i donot know how make single shareable objects

-  MockEventBus _mockEventBusClient;
-  Mokta _mokta;

**Comment by Ergamon** [comment]
Could you attach a compiling sample?

I have no clue about the classes MockEventBus, Mokta and User

**Comment by rathoregee** [comment]
Roger that

here is code repo - removing for some reason

Thank you :)

**Comment by Ergamon** [comment]
Man, that is a beer on you ;)

It took some time to get along with your code (which is beatiful, but due to the nature of complexity it still takes some time).

You start your test with:
```cs
user.Profile.Email = Email;
```

but if you take a look at SignUpAsync in SignUpService, you can see the following:
```cs
            var task1 = _oktaClient.Users.CreateUserAsync(new CreateUserWithoutCredentialsOptions
            {
                Profile = new UserProfile
                {
                    Login = signUpRequest.Email,
                    Email = signUpRequest.Email,
                    FirstName = signUpRequest.FirstName,
                    LastName = signUpRequest.LastName
                }
            });
```
So once I changed your first test line to:
```cs
signUpRequest.Email = Email;
```
This assertion passes ... and another one fails.

But this has nothing to do with AutoFixture ;)

**Comment by rathoregee** [comment]
great, i have a look into.

**Comment by rathoregee** [comment]
you have code now
if use in OktaUserSpecimenBuilder
```cs
public object Create(object request, ISpecimenContext context)
        {
            if (request is null)
            {
                throw new ArgumentNullException(nameof(request));
            }

            if (context is null)
            {
                throw new ArgumentNullException(nameof(context));
            }

            if (!RequestSpecification.IsSatisfiedBy(request))
            {
                return new NoSpecimen();
            }

            var user = context.Create<User>();

            return user;
        }
```
i get error
Message: 
```
    AutoFixture.ObjectCreationExceptionWithPath : AutoFixture was unable to create an instance of type AutoFixture.Kernel.SeededRequest because the traversed object graph contains a circular reference. Information about the circular path follows below. This is the correct behavior when a Fixture is equipped with a ThrowingRecursionBehavior, which is the default. This ensures that you are being made aware of circular references in your code. Your first reaction should be to redesign your API in order to get rid of all circular references. However, if this is not possible (most likely because parts or all of the API is delivered by a third party), you can replace this default behavior with a different behavior: on the Fixture instance, remove the ThrowingRecursionBehavior from Fixture.Behaviors, and instead add an instance of OmitOnRecursionBehavior:
    
    fixture.Behaviors.OfType<ThrowingRecursionBehavior>().ToList()
        .ForEach(b => fixture.Behaviors.Remove(b));
    fixture.Behaviors.Add(new OmitOnRecursionBehavior());
```

if i use like this
```cs
    public object Create(object request, ISpecimenContext context)
        {
            if (request is null)
            {
                throw new ArgumentNullException(nameof(request));
            }

            if (context is null)
            {
                throw new ArgumentNullException(nameof(context));
            }

            if (!RequestSpecification.IsSatisfiedBy(request))
            {
                return new NoSpecimen();
            }          

            var user = new User
            {
                ["id"] = "new-okta-id",
                Profile = new UserProfile
                {

                }
            };

            return user;
        }
```
then it works

i donot use this way

```cs
  var user = new User
            {
                ["id"] = "new-okta-id",
                Profile = new UserProfile
                {

                }
            };
```
any suggestion!

**Comment by Ergamon** [comment]
This time it is AutoFixture and more or less expected behavior. It has nothing to do with your library at all. You can create this error with any specimenbuilder written that way.

Your specimenbuilder tells AutoFixture that he likes to be responsible for creating User objects. So any time AutoFixture tries to create a User object, it calls your specimenbuilder. But your specimenbuilder calls back into AutoFixture for creation of a User object and your customized Fixture calls your specimenbuilder and so on.

Basically if all you want to do is having a User object created, just remove the specimenbuilder and everything works.

**Comment by rathoregee** [comment]
Ok - just tried 
```cs
 public static class FixtureHelper
    {
        public static IFixture CreateFixture()
        {
            var fixture = new Fixture();

            fixture.Customize(new AutoMoqCustomization
            {
                GenerateDelegates = true,
                ConfigureMembers = true
            });

            fixture.Customizations.Add(new HttpClientSpecimenBuilder());
            //fixture.Customizations.Add(new OktaUserSpecimenBuilder());
            return fixture;
        }
    }
```
now 
```cs
 [Theory, AutoMoqData]
        public async Task CreateUser([Frozen] Mokta mokta,
            [Frozen] MockEventBus mockEventBusClient,
            HttpClient client,
            User user,
            SignUpRequest signUpRequest,             
            MailAddress email, 
            string NewPersonId)
        {
```

but  profile is coming null  from user.Profile 

so it is populating Profile propetry

**Comment by rathoregee** [comment]
its working in this style
```cs
 public static class FixtureHelper
    {
        public static IFixture CreateFixture()
        {
            var fixture = new Fixture();

            fixture.Customize(new AutoMoqCustomization
            {
                GenerateDelegates = true,
                ConfigureMembers = true
            });

            fixture.Customizations.Add(new HttpClientSpecimenBuilder());           
            fixture.Inject(fixture.Build<User>()
                .With(x=>x.Id,"some-id")
                .With(x => x.Profile, new UserProfile { }).Create());
            return fixture;
        }
    }
```

**Comment by Ergamon** [comment]
Yes, but that is more or less a problem of your code and has nothing todo with AutoFixture.

The following test (without any sign of AutoFixture) fails:
```cs
    [Fact]
    public void Test()
    {
      // Given
      var profileMock = new Mock<IUserProfile>();
      var user = new User();

      // When
      user.Profile = profileMock.Object;

      // Then
      user.Profile.Should().NotBeNull();
    }
```

The source of your SDK is not included, but by disassembling your assembly, I can see the following:

```cs
    public IUserProfile Profile
    {
      get
      {
        return GetResourceProperty<UserProfile>("profile");
      }
      set
      {
        base["profile"] = value;
      }
    }
```

I guess you wanted to write `IUserProfile` not `UserProfile`.

**Comment by rathoregee** [comment]
```cs
public sealed class User : Resource, IUser, IResource
    {
        public User();
        public IUserProfile Profile { get; set; }
```

yes - i want to `IUserProfile`  property autopouplated with default value

correct me - is this correct approch?
```cs
fixture.Inject(fixture.Build<User>()
            .With(x=>x.Id,"some-id")
            .With(x => x.Profile, new UserProfile { }).Create());
```

otherwide profile will be always null in user param
```cs
[Theory, AutoMoqData]
public async Task CreateUser(User user,

{
//user.Profile always null
user.Profile.Email .....

}
```

**Comment by Ergamon** [comment]
I think you dont get what I wrote. 

Your Api suggests that the User class can have any kind of class implementing the interface IUserProfile. AutoFixture only sees the property Profile of type IUserProfile with a getter and a setter. So AutoFixture tries to create this type.

Normally AutoFixture cannot create interfaces, but due to the AutoMoqCustomization AutoFixture creates a Mock implementing this interface.

The error is in your SDK. Either the property Profile should be of type UserProfile or your code to get it back should not require your own class. What is the point of having this interface, if you dont use it?

Of course you can create a specimenbuilder responsible for IUserProfile returning a UserProfile.

But more or less the behavior of AutoFixture is logic. A property of type IUserProfile is filled with a class implementing the interface. There is no way that AutoFixture can figure out without help that you dont want any class implementing the interface, but your own.
